Signal/src/ViewControllers/OWS2FASettingsViewController.m

```diff
@@ -0,0 +1,493 @@
+                    addItem:[OWSTableItem disclosureItemWithText:
+                                              NSLocalizedString(@"ENABLE_2FA_VIEW_DISABLE_2FA",
+                                                  @"Label for the 'enable two-factor auth' item in the settings view")
+                                                     actionBlock:^{
+                                                         [weakSelf tryToDisable2FA];
+                                                     }]];
+            } else {
+                [section
+                    addItem:[OWSTableItem disclosureItemWithText:
+                                              NSLocalizedString(@"ENABLE_2FA_VIEW_ENABLE_2FA",
+                                                  @"Label for the 'enable two-factor auth' item in the settings view")
+                                                     actionBlock:^{
+                                                         [weakSelf showEnable2FAWorkUI];
+                                                     }]];
+            }
+            [contents addSection:section];
+            self.tableViewController.contents = contents;
+            break;
+        }
+        case OWS2FASettingsMode_SelectPIN:
+        case OWS2FASettingsMode_ConfirmPIN:
+            return;
+    }
+}
+
+- (BOOL)shouldHaveNextButton
+{
+    switch (self.mode) {
+        case OWS2FASettingsMode_Status:
+            return NO;
+        case OWS2FASettingsMode_SelectPIN:
+        case OWS2FASettingsMode_ConfirmPIN:
+            return [self hasValidPin];
+    }
+}
+
+- (void)updateNavigationItems
+{
+    // Note: This affects how the "back" button will look if another
+    //       view is pushed on top of this one, not how the "back"
+    //       button looks when this view is visible.
+    self.navigationItem.backBarButtonItem =
+        [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"VERIFICATION_BACK_BUTTON",
+                                                   @"button text for back button on verification view")
+                                         style:UIBarButtonItemStylePlain
+                                        target:self
+                                        action:@selector(backButtonWasPressed)];
+
+    if (self.shouldHaveNextButton) {
+        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc]
+            initWithTitle:NSLocalizedString(@"ENABLE_2FA_VIEW_NEXT_BUTTON",
+                              @"Label for the 'next' button in the 'enable two factor auth' views.")
+                    style:UIBarButtonItemStylePlain
+                   target:self
+                   action:@selector(nextButtonWasPressed)];
+    } else {
+        self.navigationItem.rightBarButtonItem = nil;
+    }
+}
+
+#pragma mark - UITextFieldDelegate
+
+- (BOOL)textField:(UITextField *)textField
+    shouldChangeCharactersInRange:(NSRange)range
+                replacementString:(NSString *)insertionText
+{
+    // TODO: ?
```

Comment:
Is the TODO here to make sure we converge with Android's constraints? Could you update the comment?

Reply:
It's around that constant which is fixed in the next PR.